site/content/docs/contributing/project-structure.md

```diff
@@ -9,42 +9,139 @@ menu:
+
+To understand this process better, we need to keep in mind that kind’s configuration schema dictates how to bootstrap a Kubernetes cluster. The schema is defined in
+[kind/pkg/cluster/config](/pkg/cluster/config).
+In this directory, currently, you will see two subdirectories:
+[v1beta1](/pkg/cluster/config/v1beta1) and
+[v1beta2](/pkg/cluster/config/v1beta2).
+Each of these subdirectories correspond to a version of kind’s cluster configuration.
+
+One of the concerns with versioned configurations is enabling the project to be compatible with all. With this in mind, [kind/pkg/cluster/config](/pkg/cluster/config) contains the main configuration fields while [v1beta1](/pkg/cluster/config/v1beta1) and [v1beta2](/pkg/cluster/config/v1beta2) build on top of it.
+
+The way this is implemented is by running deepcopy-gen and defaulter-gen in
+[kind/pkg/cluster/config](/pkg/cluster/config),
+followed by running deepcopy-gen, defaulter-gen, and conversion-gen on all version subdirectories (i.e., [v1beta1](/pkg/cluster/config/v1beta1) and [v1beta1](/pkg/cluster/config/v1beta1)).
+
+The [kubernetes/code-generator](https://github.com/kubernetes/code-generator) tools work by comment tags which are specified in the `doc.go` file for each directory. For example, all `doc.go` files within [kind/pkg/cluster/config](/pkg/cluster/config) have the following tags:
+```
+// +k8s:deepcopy-gen=package
+// +k8s:conversion-gen=sigs.k8s.io/kind/pkg/cluster/config
+// +k8s:defaulter-gen=TypeMeta
+```
+
+Additionally, [pkg/cluster/config/types.go](pkg/cluster/config/types.go) has an additional tag:
+```
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+```
+
+Let’s see how this tags work with the
+[kubernetes/code-generator](https://github.com/kubernetes/code-generator)
+binaries deepcopy-gen, defaulter-gen, and conversion-gen.
+
+For each of the directories related to defining a configuration for kind, we start by running [deepcopy-gen](https://godoc.org/k8s.io/code-generator/cmd/deepcopy-gen). deepcopy-gen generates functions that efficiently perform a full deep-copy of each type that is part of the configuration. These functions will be used to pass values to each type, go from a versioned configuration to the top level one.
+
+Once we have these utility functions in place then we will need to run
+[defaulter-gen](https://godoc.org/k8s.io/code-generator/cmd/defaulter-gen)
+to generate efficient defaulters (functions that will fill in default value for configuration fields) for the configuration schema based on the
+[`Config`](/pkg/cluster/config/types.go)
+and the [`Node`](/pkg/cluster/config/types.go) types.
+The way this works is that the
+`// +k8s:defaulter-gen=TypeMeta`
+comment tag will generate a defaulter for the `Config` type as this possesses a `TypeMeta` field.
+
+The `TypeMeta` field is defined in
```

Comment:
I'm on the fence whether this is relevant to kind developers. Whilst it *does* explain why this comment is there, I feel like it might be better placed to put this sort of comment in the code itself, a 'comment for the comment' 😅

Reply:
I want to advocate for this level of detail - for me, coming with little to no k-dev experience, it as really useful to discover that detail.